src/gov/nist/javax/sip/stack/SIPTransactionStack.java

```diff
@@ -1030,6 +1056,26 @@ public void removeDialog(String dialogId) {
         }
         dialogTable.remove(dialogId);
     }
+    
+    /**
+     * Adds a new dialog injection listener to this transaction.
+     * 
+     * @param newListener
+     *          Listener to add.
+     */
+    public void addDialogInjectionListener(SIPDialogInjectionListener newListener) {
+        dialogInjectionListeners.add(newListener);
```

Comment:
please consider using a concurrent collection better than synchronize sections (ConcurrentSkipListSet, CopyOnWriteArraySet) so thread contention is minimized. Please consider how the collection is going to be modified (listeners added/removed), and how the iterator on "find" method will be usage in terms of concurrency... (number of reads against number of writes...)